test/config_parser_SUITE.erl

```diff
@@ -1395,3 +2983,116 @@ test_equivalence_between_files(Config, File1, File2) ->
     compare_unordered_lists(lists:filter(fun filter_config/1, Opts1), Opts2,
                             fun handle_config_option/2).
+
+parse_with_host(Config) ->
+    [F] = parse(Config),
+    apply(F, [?HOST]).
+
+set_pl(K, V, List) ->
+    lists:keyreplace(K, 1, List, {K, V}).
+
+create_files(Config) ->
+    %% The files must exist for validation to pass
+    Root = small_path_helper:repo_dir(Config),
+    file:make_dir("priv"),
+    PrivkeyPath = filename:join(Root, "tools/ssl/mongooseim/privkey.pem"),
+    CertPath = filename:join(Root, "tools/ssl/mongooseim/cert.pem"),
+    CaPath = filename:join(Root, "tools/ssl/ca/cacert.pem"),
+    ok = file:write_file("priv/access_psk", ""),
+    ok = file:write_file("priv/provision_psk", ""),
+    ok = filelib:ensure_dir("www/muc/dummy"),
+    ensure_copied(CaPath, "priv/ca.pem"),
+    ensure_copied(CertPath, "priv/cert.pem"),
+    ensure_copied(PrivkeyPath, "priv/dc1.pem").
+
+ensure_copied(From, To) ->
+    case file:copy(From, To) of
+        {ok,_} ->
+            ok;
+        Other ->
+            error(#{what => ensure_copied_failed, from => From, to => To,
+                    reason => Other})
+    end.
+
+pl_merge(L1, L2) ->
+    M1 = maps:from_list(L1),
+    M2 = maps:from_list(L2),
+    maps:to_list(maps:merge(M1, M2)).
+
+%% Runs check_one_opts, but only for fields, that present in both
+%% MongooseIM and TOML config formats with the same name.
+%% Helps to filter out riak fields automatically.
+check_one_opts_with_same_field_name(M, MBase, Base, T) ->
+    KeysM = maps:keys(maps:from_list(MBase)),
+    KeysT = lists:map(fun b2a/1, maps:keys(Base)),
+    Keys = ordsets:intersection(ordsets:from_list(KeysT),
+                                ordsets:from_list(KeysM)),
+    check_one_opts(M, MBase, Base, T, Keys).
+
+check_one_opts(M, MBase, Base, T) ->
+    Keys = maps:keys(maps:from_list(MBase)),
+    check_one_opts(M, MBase, Base, T, Keys).
+
+check_one_opts(M, MBase, Base, T, Keys) ->
+    [check_one_opts_key(M, K, MBase, Base, T) || K <- Keys].
+
+check_one_opts_key(M, K, MBase, Base, T) when is_atom(M), is_atom(K) ->
+    BK = atom_to_binary(K, utf8),
+    MimValue = maps:get(K, maps:from_list(MBase)),
+    TomValue = maps:get(BK, Base),
+    Mim = [{K, MimValue}],
+    Toml = #{BK => TomValue},
+    ?_eqf(modopts(M, Mim), T(Toml)).
+
+binaries_to_atoms(Bins) ->
+    [binary_to_atom(B, utf8) || B <- Bins].
+
+run_multi(Cases) ->
+    Results = [run_case(F) || {F,_} <- Cases],
+    case lists:all(fun(X) -> X =:= ok end, Results) of
+        true ->
+            ok;
+        false ->
+            Failed = [Zip || {Res,_}=Zip <- lists:zip(Results, Cases), Res =/= ok],
+            [ct:pal("Info: ~p~nResult: ~p~n", [Info, Res]) || {Res, Info} <- Failed],
+            ct:fail(#{what => run_multi_failed, failed_cases => length(Failed)})
+    end.
+
+run_case(F) ->
+    try
+        F(), ok
+    catch Class:Reason:Stacktrace ->
+        {Class, Reason, Stacktrace}
+    end.
```

Comment:
Quite a bit of complexity here :-(
I guess I would not ask for any rewrite here (change the granularity, use the actual CT test cases with, when needed, CTH), that would be cruel, let's keep it.